### ISSUE TITLE
Feature Discussion: Custom `Err` types in `Result`

### ISSUE THREAD

**Issue by OliverBrotchie** [OliverBrotchie/optionals#8]
### Original Comment

This function could produce an Ok.

```TypeScript
const a = 2 as unknown as Error
const res = Err(a) // Not a string, not an Error, but TypeScript thinks so
console.log(res.isOk() // true
console.log(res.isErr() // false
```

Perhaps a generic `isErrorLike(input: any): input is Error` function could be created which checks for characteristics of objects like Errors could be used in this function as well as the `Result.isErr` function.

```TypeScript
interface ErrorKind {
	message: string;
}
type ErrorLike = Error | ErrorKind;

function isErrorLike(input: any): input is ErrorLike {
	return (
		input instanceof Error ||
		(input &&
			typeof input === 'object' &&
			(Object.prototype.isPrototypeOf.call(Error, input) ||
				typeof input.message === 'string'))
	);
}
```
I am kind of concerned about forcing Error as the only type here, as then we get undefined behavior at runtime anyway... should it throw if anything not Error or string is passed?

I think Rust allows any value to be passed for E in Result<T, E>, and JS allows us to throw any value, so perhaps there could be some other signal that what's constructed is an error?

Perhaps an optional `type: 'Err' | 'Ok' = 'Ok'` as an arg, and then always pass `'Err'` from `function Err`?

Such a type value could then also be used in the `Result.isOk` and `Result.isErr` functions to avoid prototype chain lookups which are currently used to decide `Result.isErr`. Might provide a performance bump, and it's how I do it in my private lib.

_Originally posted by @aaronhuggins in https://github.com/OliverBrotchie/optionals/pull/7#discussion_r1008324662_

**Comment by OliverBrotchie** [comment]
> ```ts
>const a = 2 as unknown as Error
>const res = Err(a) // Not a string, not an Error, but TypeScript thinks so
>console.log(res.isOk() // true
>console.log(res.isErr() // false
>```

I don't think we should worry about someone performing incorrect type casting - `as` is something that TS developers should try to avoid.

> Perhaps a generic isErrorLike(input: any): input is Error function could be created which checks for characteristics of objects like Errors could be used in this function as well as the Result.isErr function.
>```ts
>interface ErrorKind {
>	message: string;
>}
>
>type ErrorLike = Error | ErrorKind;
>
>function isErrorLike(input: any): input is ErrorLike {
>	return (
>		input instanceof Error ||
>		(input &&
>			typeof input === 'object' &&
>			(Object.prototype.isPrototypeOf.call(Error, input) ||
>				typeof input.message === 'string'))
>	);
>}
>```
>I am kind of concerned about forcing Error as the only type here, as then we get undefined behavior at runtime anyway... should it throw if anything not Error or string is passed?
>
>I think Rust allows any value to be passed for E in Result<T, E>, and JS allows us to throw any value, so perhaps there could be some other signal that what's constructed is an error?
>
>Perhaps an optional type: 'Err' | 'Ok' = 'Ok' as an arg, and then always pass 'Err' from function Err?
>
>Such a type value could then also be used in the Result.isOk and Result.isErr functions to avoid prototype chain >lookups which are currently used to decide Result.isErr. Might provide a performance bump, and it's how I do it in >my private lib.

My thoughts at the time of sticking to `Error` is that it is the native JS way of error handling so it allows for a bit more interop and it provides the unique benefit of a call stack. 

You are definitely correct that other values should be returnable in Err, the question is how restrictive should it be? I think we should 100% at the very least allow `enum`s to be passed.

**Comment by aaronhuggins** [comment]
Idea, going with your point about interop. Perhaps we add an error type that catches all non-ErrorLike, non-string values?

The reason I bring this up is that consumers of `Err` may have cases where there's little control over the *exact* value that's passed or produced, resulting in *runtime* ambiguity. Not even speaking to typecasting, just an inherent footgun in JavaScript.

Example code:
```TypeScript
// amiguous_error.ts
export class AmbiguousError extends Error {
  value: any
  constructor (value: any) {
    super(`Error value passed to Result<T, E> was ambiguous: ${Object.prototype.toString.call(value)}`)
    this.value = value
    this.name = 'AmbiguousError'
    Object.setPrototypeOf(this, AmbiguousError.prototype)
  }
  
  toJSON () {
    return {
      message: this.message,
      name: this.name,
      stack: this.stack,
      value: this.value
    }
  }
}

// result.ts
export function Err ... {
  if (typeof error === 'string') ...
  if (isErrorLike(error) ...
  else {
    return new Result(new AmbiguousError(error))
  }
}
```

**Comment by aaronhuggins** [comment]
FYI, I'm a sucker for JSON serializable errors when given the chance.

**Comment by aaronhuggins** [comment]
I'd probably ditch `interface ErrorKind` and `type ErrorLike` and just check for `Error` if the `AmbiguousError` class were implemented.

```TypeScript
function isError(input: any): input is Error {
  return (
    input instanceof Error ||
    (input &&
      typeof input === 'object' &&
      Object.prototype.isPrototypeOf.call(Error, input))
    );
}
```

**Comment by OliverBrotchie** [comment]
> Idea, going with your point about interop. Perhaps we add an error type that catches all non-ErrorLike, non-string values?
>
>The reason I bring this up is that consumers of Err may have cases where there's little control over the exact value that's passed or produced, resulting in runtime ambiguity. Not even speaking to typecasting, just an inherent footgun in JavaScript.

I totally agree! Let's make sure that the Result either extends Err (`Erro`r and `AmbiguousError`) or possibly an `enum` if we can find a simple way of doing that.

**Comment by OliverBrotchie** [comment]
> FYI, I'm a sucker for JSON serializable errors when given the chance.

I am too! 😆

This does however preclude `AmbiguousError` from `implement`'ing `Error`.

**Comment by OliverBrotchie** [comment]
Closing as won't fix.

Custom error types are already fully supported: extend `Error`.

```ts
class NotFoundError extends Error {}
class RateLimitError extends Error { constructor(msg: string, readonly retryAfter: number) { super(msg); } }

Err(new NotFoundError("x"))        // works today
Err(new RateLimitError("y", 60))    // works today, carries retryAfter
```

This gives type safety (`E extends Error`), stack traces, and cross-realm/custom-subclass detection. The only uncovered case is arbitrary non-`Error` values (`Err(42)`, `Err({ code })`), which is Rust's 'any E' — niche, not requested since 2022, and would require dropping the `Error` discriminator that powers `isErr()`. Reopen if a real use case needs non-Error error values.
